File: 6/6.23.swift
// 6.23
// [GOOD]
// プロトコルに準拠したclassやstructなどは
// プロトコルで定義されている変数やメソッドは
// 絶対実装しなければいけない「約束事」となる

// つまりこのShapeプロトコルに準拠したらareaは絶対ある
protocol Shape {
    var area: Double { get }
}

struct Rectangle: Shape {
    let width: Double
    let height: Double
    
    // RectangleはShapeの約束事であるareaを実装しないといけない
    var area: Double {
        width * height
    }
}


struct Circle: Shape {
    let radius: Double
    
    init(radius: Double) {
        self.radius = radius
    }
    
    // CircleはShapeの約束事であるareaを実装しないといけない
    var area: Double {
        radius * radius * Double.pi
    }
}

// 6.21
// [Good]
// Shapeの約束を守っているstruct(classなど)ならなんでも代入できる
// そのため先ほどのshowAreaメソッドはいらないし型判定の分岐もしなくていい
let shape: Shape = Circle(radius: 10)
print(shape.area)


// [おまけ]
// これはあくまでprotocolの「ひとつの使い方」の話であり
// protocolを使えば他にも色んな使い方がある
// iOSなら定番のdelegateパターンもprotocolのひとつの使い方

File: 6/6.23.dart
// [GOOD]
// アブストラクトクラスで変数やメソッドを定義すれば
// それに準拠したクラスではそれらを必ず実装しなければならない「約束事」となる
abstract class Shape {
  double get area;
}

class Rectangle extends Shape {
  final double width;
  final double height;

  Rectangle({required this.width, required this.height});

  // RectangleはShapeの約束事であるareaを実装しないといけない
  @override
  double get area => width * height;
}


class Circle extends Shape {
  final double radius;

  Circle({required this.radius});

  // CircleはShapeの約束事であるareaを実装しないといけない
  @override
  double get area => radius * radius * 3.14159265359;
}

void main() {
  // [Good]
  // Shapeの約束を守っているクラスならなんでも代入できる
  // そのため先ほどのshowAreaメソッドはいらないし型判定の分岐もしなくていい
  Shape shape = Circle(radius: 10);
  print(shape.area);
}

// [おまけ]
// これはあくまでアブストラクトクラスの「ひとつの使い方」の話であり
// アブストラクトクラスを使えば他にも色んな使い方がある
// Flutterなら定番のcallbackパターンもアブストラクトクラスのひとつの使い方
